### ISSUE TITLE
DPMultiheadAttention Doesn't check attention mask correctly when batch_first is set to True.

### ISSUE THREAD

**Issue by Ghazi-Echo** [pytorch/opacus#774]
## 🐛 Bug

This bug happens when using `DPMultiheadAttention` with the `batch_first` parameter set to True. While going thruough the shape checks, the forward method of `DPMultiheadAttention` check whether the provided mask two last dimensions are `(query.shape(0), key.shape(0))`. This is only correct when batch_first=False.

## To Reproduce
[COLLAB LINK](https://colab.research.google.com/drive/19HriI3H1_k4P8JrcYax2L3Z-ckWp7uI-?usp=sharing)

1. Build a model with a Transformer Encoder that uses `DPMultiheadAttention` as a replacement for vanilla multi-head attention and the batch_first parameter set to True.
2. Make a forward pass with an attention mask.
3. Watch the forward pass fail on the line where it checks the coherence between attention mask shape and query/key first dimension size.

Check the collab example above where I reproduce the error, and show that a patch solves it.

## Expected behavior
I expect the mask size to be matched to the dimension 0 of query/key when batch_first=False and to dimension1  of query/key when batch_first=True.

## Environment

PyTorch version: 2.7.0+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.11.11 (main, Dec  4 2024, 08:55:07) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.6.87.2-microsoft-standard-WSL2-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        39 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               16
On-line CPU(s) list:                  0-15
Vendor ID:                            GenuineIntel
Model name:                           12th Gen Intel(R) Core(TM) i7-12650H
CPU family:                           6
Model:                                154
Thread(s) per core:                   2
Core(s) per socket:                   8
Socket(s):                            1
Stepping:                             3
BogoMIPS:                             5376.00
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology tsc_reliable nonstop_tsc cpuid tsc_known_freq pni pclmulqdq vmx ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves avx_vnni vnmi umip waitpkg gfni vaes vpclmulqdq rdpid movdiri movdir64b fsrm md_clear serialize flush_l1d arch_capabilities
Virtualization:                       VT-x
Hypervisor vendor:                    Microsoft
Virtualization type:                  full
L1d cache:                            384 KiB (8 instances)
L1i cache:                            256 KiB (8 instances)
L2 cache:                             10 MiB (8 instances)
L3 cache:                             24 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-15
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Vulnerable: No microcode
Vulnerability Retbleed:               Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] mypy_extensions==1.1.0
[pip3] numpy==1.26.4
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.77
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.5.1.17
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.6.3
[pip3] nvidia-nccl-cu12==2.26.2
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] torch==2.7.0
[pip3] triton==3.3.0
[conda] Could not collect

**Comment by immu4989** [comment]
Looking at [opacus/layers/dp_multihead_attention.py](https://github.com/meta-pytorch/opacus/blob/main/opacus/layers/dp_multihead_attention.py) on main, the attn_mask shape check still uses query.size(0) / key.size(0) unconditionally:

```python
if attn_mask.dim() == 2:
attn_mask = attn_mask.unsqueeze(0)
if list(attn_mask.size()) != [1, query.size(0), key.size(0)]:
raise ValueError("The size of the 2D attn_mask is not correct.")
elif attn_mask.dim() == 3:
if list(attn_mask.size()) != [
bsz * self.num_heads,
query.size(0),
key.size(0),
]:
raise ValueError("The size of the 3D attn_mask is not correct.")
```

A few lines above, tgt_len is already computed correctly for both batch_first paths — the validation just doesn't use it. The straightforward fix is to introduce a parallel src_len and reference both:

```python
if not self.batch_first:
tgt_len, bsz, embed_dim = query.size()
src_len = key.size(0)
else:
bsz, tgt_len, embed_dim = query.size()
src_len = key.size(1)
...
if list(attn_mask.size()) != [1, tgt_len, src_len]:
raise ValueError("The size of the 2D attn_mask is not correct.")
```

@ayacherigui's size(1) workaround works for batch_first=True but would silently break the batch_first=False path, so I don't think a flat size(0) → size(1) swap is the right fix.

@iden-kalemaj — is there still an internal fix queued for this, or is it open for a community PR? Happy to send one with a batch_first regression test covering both the 2D and 3D mask paths. The DP guarantee surface isn't touched — this is purely input-shape validation.

**Comment by iden-kalemaj** [comment]
@immu4989 You are welcome to submit a PR. Ideally, we would also expand the existing unit tests [here](opacus/tests/dp_layers/dp_multihead_attention_test.py) to include both cases where batch_size=True and when batch_size=False.

**Comment by immu4989** [comment]
Heads up , adding the batch_first test coverage you asked for surfaced a separate pre existing bug. With batch_first=True (no mask, no add_bias_kv, no add_zero_attn), `DPMultiheadAttention` output diverges significantly from `nn.MultiheadAttention`:

```
batch_first=False: max|diff|=0      (norms: 1.8235 vs 1.8235)
batch_first=True:  max|diff|=0.961  (norms: 1.8235 vs 1.8165)
```

Norms are within ~0.7% but max|diff| is on the same scale as the values themselves — signature of elements being rearranged, not corrupted. Looking at [dp_multihead_attention.py:354-359](https://github.com/meta-pytorch/opacus/blob/main/opacus/layers/dp_multihead_attention.py#L354-L359):

```python
if self.batch_first:
attn_output = attn_output.contiguous().view(bsz, tgt_len, embed_dim)
else:
attn_output = (
attn_output.transpose(0, 1).contiguous().view(tgt_len, bsz, embed_dim)
)
```

The `batch_first=False` branch correctly handles the layout via `transpose(0, 1) → contiguous → view`. The `batch_first=True` branch skips the transpose and reinterprets `(bsz * num_heads, tgt_len, head_dim)` directly as `(bsz, tgt_len, embed_dim)` — so head positions get spliced into time positions. Existing tests never caught this because they all use the default `batch_first=False`.

@iden-kalemaj — three options, would prefer your call:

1. Bundle both fixes in this PR (they're in the same forward path).
2. Land mask validation here, separate issue + PR for the reshape.
3. Narrow the test to assert only that the mask is accepted under `batch_first=True` (no output-parity check), defer the reshape question.

Slight preference for (1) since the test coverage you asked for is what surfaced it — happy to do whichever fits your review workflow best.

**Comment by iden-kalemaj** [comment]
@immu4989 thank you for surfacing this, agreed that 1 is preferred. 

If you run into obstacles with fixing the reshape issue, you can first just land the change for mask validation with a narrower test.
